repos/system_upgrade/common/libraries/rpms.py

```diff
@@ -114,3 +145,41 @@ def check_file_modification(config):
     """
     output = _read_rpm_modifications(config)
     return _parse_config_modification(output, config)
+
+
+def get_leapp_packages(source_version=None, component=None, include_deps=False):
+    """
+    Get list of leapp packages.
+
+    :param source_version: a list or string specifying source_versions. If not defined then all packages
+        for current source_target_version will be returned.
+    :param component: a list or string specifying leapp components. If defined then only packages related
+        to the specific component will be returned.
+    :param include_deps: a flag to control whether deps meta packages are to be filtered out.
+    :raises ValueError: if a requested component or source_version doesn't exist.
+    """
+    res = set()
+    source_versions = [source_version] if isinstance(source_version, str) else source_version
+    if not source_versions:
+        # No source_version of interest specified -> treat as if only current source system version
+        # requested
+        source_versions = [get_source_major_version()]
+    components = [component] if isinstance(component, str) else component
+    if not component:
+        # No component of interest specified -> return default leapp components
+        components = _LEAPP_DEFAULT_COMPONENTS
+    for comp in components:
+        for src_version in source_versions:
+            if comp not in _LEAPP_PACKAGES_MAP:
+                error_msg = "The requested component {comp} is unknown, available choices are {choices}".format(
+                        comp=component, choices=sorted(_LEAPP_PACKAGES_MAP.keys()))
+                raise ValueError(error_msg)
+            if src_version not in _LEAPP_PACKAGES_MAP[comp]:
+                error_msg = "The requested source_version {ver} is unknown, available choices are {choices}".format(
+                        ver=src_version, choices=sorted(_LEAPP_PACKAGES_MAP[comp].keys()))
+                raise ValueError(error_msg)
```

Comment:
I would like to know your opinion(s) on that guys. I am thinking in this way:
* we are responsible that the src_version is defined same for all `Leapp.Components`. 
* If some new component is added, or some existing dropped, it should return empty list.
* If I create a generic actor, that will use this function for all upgrade paths, and we realize it will crash under specific conditions for a new major upgrade path, it could be still too late (even unit-tests will not count with that as they will be written based on the situation that exists during writting the code calling this function).

The implementation does not need to be changed, but in case we have same understanding / opinion on the scenariou I mentioned above, it would make sense to put this check before the loop, to check just the generic list of possible values. In this case `['7', '8']` that will be soon extended by `'9'`. WDYT? @oamg/developers 

To be clear, this one is not blocker, but it's better to point this out so people think about that.

Reply:
Sorrt, but I am against overcomplicating the logic at this point and committing ourselves to the Leapp.Components support cycle. The goal of this pr is to introduce a library helper - nothing more and nothing less.

Reply:
once this is part of the library, it's already supported by us.

Reply:
Sorry, I am completely missing what you are suggesting here, I see only many ifs and fear of something failing that is not even there yet.

If something crashes we fix it. The api doesn't change either way. What are we discussing here?

Reply:
Sorry, I tried to provide it the example in the description of this thread. But I will try it in a different way. So e.g.:
```
_MY_LEAPP_PACKAGES_MAP = {
    "framework": {'7': {'pkgs': ['leapp', 'python2-leapp'],
                                          'deps': ['leapp-deps']},
                                    '8': {'pkgs': ['leapp', 'python3-leapp'],
                                          'deps': ['leapp-deps']}
                                    },
    "newcomponent":  {'8': {'pkgs': ['leapp', 'python3-leapp'],
                                          'deps': ['leapp-deps']}
                                    },
```
calling `get_leapp_packages("7", "newcomponent")`  ends with an error now. As mentioned in the start of this thread, I will point out one more time key things:
* I think we should in such a case return empty list
* valid values in such a scenario are "7", "8".
* such a scenario should not happen, and when the map is changed in any way, all releases should be always defined. so in case of "newcomponent" should be always defined "7" & "8" in this scenario
  * it's simple to cover such a scenario in the code, where having undefined release lead to an empty list

Why I started the thread is, that if we agree on that, we can modify the implementation a little bit, to make it more clear and safe. Which would mean putting one if statement out of the loop, and in a case we would like to keep some releases "undefined", we could add one `if` into the loop to automatically act like an empty list is defined. pseudocode:
```
def get_leapp_packages(...):
    ...
    if major_release and major_release not in ["7", "8", ..]:
       raise ValueError
   ...
   for comp in components:
       if release not in _MAP[comp]:
           # nothing to do - release is valid, but undefined for the particular component
           log.warning(...) 
           continue
```

Reply:
Why should the behavior be different for `get_leapp_component('7', 'nosuchcomponent')` from raising ValueError as we agreed before? IIRC we had this changed specifically because we wanted to provide clear fast failure for anyone who uses it wrong, and the empty list doesn't help much here.

Reply:
@fernflower  because for me it's different whether I put call `get_leapp_component("1")` - with completely nonsense value for a release that we do not cover anywhere in the code / that is unsupported at all, `and sitaution when `get_leapp_component("7")` is valid but our mistake that the map is not full. As I wrote, we can take full responsibility, so with any change, the author and reviewer will ensure the map will be always complete. But I raise that to think about that, that we could prevent us from issues. As I wrote, it's not blocker,  I just started the public discussion to get people' opinions

Reply:
Okay, my position stands - I have no strong feelings about ValueError (crash early in case of wrong usage) or empty list (not getting a value and possibly messing up reports), but I am strictly against a mix of both, as it has the worst of both worlds unfortunately.

Reply:
Thanks for your opinion. As I mentioned, this is not blocker and it can be changed in future in case we will want as now there is not yet an existing scenario in the code for the behaviour we are discussing. Fix please the other stuff as we discsused so we could finish this PR. Keeping this thread opened for others.

Reply:
I think we should raise `ValueError` in case there is an unknown component, but your point about making it easier for people to write generic actors is valid.  I think we could solve it in another way.  However, I can also see two scenarios: One in favour of making it easier and one in favour of forcing the caller to update their code:

### In favour of making it easier:
* `leapp-cockpit` needs to be factored into the RHEL 7 and 8 upgrade.
* `leapp-cockpit` must not be factored into the RHEL 9 upgrade (perhaps the package is not shipped on that distro version).
* Calling code had: `get_leapp_packages(components=(LeappComponents.FRAMEWORK, LeappComponents.COCKPIT))`

In this case, if we did not error due to the use of `LeappComponents.COCKPIT`, there would be no problem and the caller would appreciate not having to modify their code to handle both 7&8 upgrades and 9 upgrades.

### In favour of forcing an update:
* `leapp-cockpit` for the RHEL 9 upgrade, `leapp-cockpit` and `snactor` must be installed together.  Therefore, `leapp-cockpit`  is moved into the `TOOLS` component. 
* Calling code had: `get_leapp_packages(components=(LeappComponents.FRAMEWORK, LeappComponents.COCKPIT))`

In this case, we need to force the caller to update their code.  If they do not get an error, they will not get both `leapp-cockpit` and `snactor` returned as leapp_packages which will be a bug in their code.

### Solution that handles both (but may be overly complex):

* For the first use case, `LEAPP_DEFAULT_COMPONENTS` can continue to have an entry for `COCKPIT` but it will be an empty list.  Then, there will be no error in the calling code.
* For the second use case, `COCKPIT` must be removed.  That way callers will be forced to update their code to do the right thing.

The danger in having two different ways of handling this as just an entry or lack of entry in the mapping is that someone might not realize that a particular change should follow one or the other strategy and introduce a bug by doing the wrong thing (this could even be a regression if someone thought they were fixing a bug in another PR but in reality, it was doing the right thing by raising an exception/not raising an exception initially).

### How to make sure that testing catches errors?

We can write a unittest that will catch whether the code will throw errors or not throw errors when we don't expect it by looping through the components listed in the mapping and making sure that ones we do not want to throw errors exist in all releases.  However, as noted in the previous section, this test is only as good as we are in making sure the test is correct.  If the test failure doesn't explain the issue well enough to the developer, they will just change the test instead of realizing their code change is a mistake.